[2단계 - 장바구니 협업] 후이(최광희) 미션 제출합니다.

kwannee opened the pull request:
안녕하세요 아사님! 후이입니다! 어느덧 마지막 피드백을 받게 됐네요… 마지막을 아사님과 함께 하게 돼 정말 영광입니다 하하

step2는 백엔드와 본격적인 협업을 했습니다. 백엔드가 만든 서버를 연동해서 제대로 구동 하는지 확인하는 작업 등을 하는데 생각보다 재밌었습니다. 그냥 msw로 할 때는 별 생각이 없었는데 실제 서버를 연동했다고 생각하니 감동이 있더라고요. 뭔가 살아 있는 생명체를 만든 것 같기도 하고 기분이 묘했습니다 허허… 

### 특이사항

- 이번 미션은 배포를 하지 못해 로컬에서만 실행이 가능합니다. 번거롭지만 로컬에서 구동 부탁드리겠습니다 ㅠㅠ
    
    ```jsx
    git clone https://github.com/kwannee/react-shopping-cart-prod.git
    cd react-shopping-cart
    npm i
    npm run start
    ```
    
- constants/api.js 에 들어가시면 서버와 연동 관련 로직이 있습니다.
    
    ```jsx
    export const test = true;
    
    const eve = 'http://ec2-54-180-96-9.ap-northeast-2.compute.amazonaws.com:8080';
    const leo = 'http://ec2-15-164-103-250.ap-northeast-2.compute.amazonaws.com:8080';
    const judy = 'http://ec2-43-200-5-8.ap-northeast-2.compute.amazonaws.com:8080';
    const bunny = 'http://ec2-3-34-179-158.ap-northeast-2.compute.amazonaws.com:8080';
    export const BASE_URL = test ? eve : 'http://localhost:3000';
    
    const CUSTOMERS = 'customers';
    export const API_URL_PATH = {
      PRODUCTS: `/products`,
      CUSTOMERS: `/${CUSTOMERS}`,
      LOGIN: `/auth/login`,
      EMAIL: `/${CUSTOMERS}/email`,
      NAME: `/${CUSTOMERS}/name`,
      CARTS: `/${CUSTOMERS}/carts`,
      EMAIL_VALIDATE: `/${CUSTOMERS}/email/validate`,
    };
    ```
    
    test를 true로 하면 서버와 연동이 되고, false로 하시면 MSW로 됩니다.
    
    `BASE_URL = test ? eve : 'http://localhost:3000';` 이 부분에서 eve를 다른 크루의 서버로 바꿀 수 있습니다. 현재 사용가능한 서버는 `eve`와 `bunny`입니다.
    
    .env 파일을 만들고 script로 각각 다른 서버를 구동하려고 해봤으나, 각 크루들의 서버를 테스트할 때마다 다시 빌드해야 하는 번거로움이 있어 차선책으로 이런 방식을 사용했습니다. 또한 아사님이 리뷰하실 때 .env 파일을 만드시는 번거로움을 겪지 않게 하려는 의도도 있었습니다.(핑계)
    

### step1 피드백

**step1 때 받았던 피드백에 대해 반영한 점을 말씀 드리겠습니다.**

1. step1 필수 구현 사항 중 제대로 동작하지 않던 것 들을 수정했습니다.
    
    시간에 쫓겨 제대로 구현하지 못 하고 제출했던 것들이 많은데요. 수정한 것은 다음과 같습니다.
    
    - 회원 가입 폼을 다 입력하지 않아도 가입이 되던 것
    - 회원 수정이 안되던 것
    - 회원 탈퇴 기능이 없던 것
    - 입력에 대해 입력 폼 위치가 움직이던 것
2. alert를 스낵바로 수정했습니다. 아사님 말대로 사용자 입장에서 거슬릴 수 있을 것 같았어요. 스낵바를 통해 유저 친화적으로 메세지를 보여줄 수 있게 했습니다. [alert 리뷰](https://github.com/woowacourse/react-shopping-cart-prod/pull/30#discussion_r889165620)
    
3. 에러 페이지에서 Home으로 이동할 수 있는 Link를 만들었습니다. [에러페이지 리뷰](https://github.com/woowacourse/react-shopping-cart-prod/pull/30#discussion_r889190941)
4. 중첩 라우팅을 사용하여, 인증된 유저만 갈 수 있는 페이지와 그렇지 않은 페이지를 구분했습니다. [중첩 라우터 리뷰](https://github.com/woowacourse/react-shopping-cart-prod/pull/30#discussion_r889188746)
    

**반영하지 못했던 것 or 궁금한 점**

1. palette 사용법.
    
    피드백 주신 애매하게 잡힌 palette에 대해 사실 피드백 적용을 하지 못했습니다. 솔직하게 말하면 게을러서 이긴 하지만, 변명을 좀 하자면 제 코드가 아니라서 이미 많은 부분들이 그렇게 돼 있었기 때문에 통일성을 위한 점도 있었습니다… 너무 변명이네요…
    
    그리고 사용법에 대해 궁금한 점도 있습니다!
    
    링크 주신 사이트 진짜 좋은 것 같고 레벨 3 때 유용하게 쓸 수 있을 것 같아 감사드립니다! 하나 궁금한 점은 
        
    <img width="992" alt="image" src="https://user-images.githubusercontent.com/41886825/173082156-1cfba56d-1db9-4eb3-9ece-3bbb1e3e5536.png">

    
    저 중에 내가 원하는 색이 500 하나라고 하더라도 모든 색을 다 저장해야 하는 걸까요?
    
    제 생각으로는 사용 되지 않는 색이라 할지라도 일단은 모두 저장 해놓는 편이 좋을 것 같긴 합니다. 버튼 hover 시 배경색을 비슷하지만 채도만 다른 색으로 변경할 때 사용하는 등 디자인적으로 일관성을 줄 수 있을 것 같다는 생각도 들고요. 한편으로는 디자이너가 다 정해주기도 하나? 하는 생각도 듭니다. 어떻게 사용하면 좋을까요?  [palette 리뷰](https://github.com/woowacourse/react-shopping-cart-prod/pull/30#discussion_r889171229)
    
2. 중첩 라우터 사용
    
    말씀해주신 중첩 라우터를 사용하여 정말 편하게 유저용, 비유저용 페이지를 구분할 수 있었습니다!
    
    하나 궁금한 점은 피드백 주신 예 중에서 Header가 포함된 페이지와 포함되지 않은 페이지를 구분할 수 있다고 하셨는데요.
    
    ```jsx
    <Routes>
      <Route path="/" element={<ProductList />} />
      <Route element={<AuthOnly />}>
        <Route path="/cart" element={<ShoppingCartList />} />
        <Route path="/user/modify" element={<ModifyUserInfo />} />
      </Route>
      <Route element={<UnAuthOnly />}>
        <Route path="/login" element={<Login />} />
        <Route path="/signup" element={<SignUp />} />
      <Route>
      <Route path="/*" element={<ErrorPage>존재하지 않는 페이지입니다</ErrorPage>} />
    </Routes>
    ```
    
    현재 코드가 이렇습니다. 만약 이 중 <ProductList />, <ShoppingCartList /> 두 컴포넌트가 Header를 사용한다고 하면 어떻게 사용해야 좋을까요? 두 컴포넌트는 `모든 유저 접근 가능` 그리고 `로그인 한 유저만 접근 가능` 으로 서로 다른 중첩 라우터로 분리 돼 있습니다.
    
    제가 생각한 방식으로는
    
    ```jsx
    <Routes>
      <Route element={<HeaderLayout />}> 
        <Route path="/" element={<ProductList />} />
      </Route> 
      <Route element={<AuthOnly />}>
        <Route element={<HeaderLayout />}> 
          <Route path="/cart" element={<ShoppingCartList />} />
        </Route> 
        <Route path="/user/modify" element={<ModifyUserInfo />} />
      </Route>
      <Route element={<UnAuthOnly />}>
        <Route path="/login" element={<Login />} />
        <Route path="/signup" element={<SignUp />} />
      <Route>
      <Route path="/*" element={<ErrorPage>존재하지 않는 페이지입니다</ErrorPage>} />
    </Routes>
    ```
    
    이런 식으로 Header가 필요한 Route에 한해 중첩으로 감싸주면 될 것 같은데요. 근데 이게 좀 지저분해 보일 수도 있겠다는 생각이 듭니다. 어떤 식으로 하면 좋을까요?
    
    [중첩 라우터 Header](https://github.com/woowacourse/react-shopping-cart-prod/pull/30#discussion_r889188746)
    

### 질문사항

1. 서버에서 오는 데이터를 사용할 수 있는 형태로 가공하는 과정이 원래 일반적인건가요?
    
    이번 협업을 하며 바로 사용할 수 있는 형태가 아닌 데이터를 서버에서 주기도 한다는 것을 처음 알았습니다.
    
    ```jsx
    carts : [ 
     {
      product:
      {
        id : 1,
        thumbnail : "https://image.istarbucks.co.kr/upload/store/skuimg/2021/02/[9200000001636]_20210225093600536.jpg",
        name : "콜드 브루 몰트",
        price : 4800,
      }
      quantity : 1
     },
    ]
    ```
    
    이런 형태인데요. 우리 입장에서는 저 product라는 객체를 한번 더 감싸지 않고 모두 같은 depth로 내려준다면 바로 쓸 수 있을 것 같아 굉장히 편할 것 같은데 백엔드는 그렇게 해주지 않더라고요. dto라는 것 때문에 그렇다는 말을 들었는데, 그렇기 때문에 저렇게 서버에서 온 데이터를 한번 가공하는 과정을 추가하는 것이 불가피해졌습니다. 그런 과정을 거치면서 `이게 일반적으로 이렇게 하는 건가?` 하는 생각이 들었습니다. 일반적으로 이렇게 하는 건가요?
    

### 마치며

긴 글 읽어주셔서 감사합니다! 날카로우면서도 부드럽고 달달한 피드백 부탁드립니다!!!

kwannee committed:
refactor: API_URL_PATH 상수화

kwannee committed:
refactor: Logo 컴포넌트 사용하는 곳에서 margin 주도록 수정

kwannee committed:
fix: 회원가입 창 입력에 따라 input height 움직이던 것 고정

kwannee committed:
fix: 회원가입 페이지 들어오자 마다 fetch하던 오류 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: useFetch에서 성공 시 값을 바로 리턴하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: cors 에러 방지 위해 Access-Control-Allow-Origin 추가

- 방지가 될까?

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: api url path 변경에 따른 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 로그아웃 안되는 문제 해결

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: accessToken이 있을 때만 이름 응답하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: 유저 정보 수정 시 정보 제대로 불러오지 않던 오류 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 유저 정보 중 phoneNumber, 서버 저장 키에 맞게 phone으로 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: logout시 accessToken을 제대로 reset하지 않던 오류 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: email 중복 체크하는 로직 간소화

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 메인 페이지에서 장바구니 상품 수량 변경 구현

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 장바구니에 담긴 수량 미리보기 추가

kwannee committed:
fix: 상품 목록 반응형으로 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: api 명세 변경에 따른 코드 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 체크된 장바구니 상품 총 금액 계산 로직 리덕스로 이동

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: ModifyQuantityBox에 0이면 쓰레기통 보이도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 장바구니에 담기면 상품 목록 다시 불러오도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: api 명세 변경에 따른 데이터 형태 변경

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: 회원 정보 모두 입력 해야 회원가입 할 수 있도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 헤더 로고 재사용하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 회원 탈퇴 기능 구현

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: shoppingCartReducer 삭제

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 로그아웃하면 redux 초기화 하고 reload

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: api 명세 변경에 따른 authorization 형태 변경

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 장바구니 개수가 1개일 때 쓰레기통 icon 보여주고 클릭 시 delete 하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: 선택된 상품 total price 로컬에서 관리하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: 상품 목록 페이지에서 장바구니 수량 변경 박스 500ms에 꺼지던 것 1500ms에 꺼지도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 장바구니에서 수량 변경 리덕스로 하던 것 api로 처리하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: 로그인 에러 메세지 사용자 친화적으로 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: api 명세 변경에 따른 header 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: MSW 장바구니 상품 삭제 로직 구현

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: useFetch에서 error throw 하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 상품 목록 불러오던 로직 커스텀 훅으로 분리

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: loadCarts 로직 커스텀 훅으로 분리

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: deleteCarts 로직 커스텀 훅으로 분리

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 장바구니 등록, 수량 수정 커스텀 훅으로 분리

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 유저 삭제 기능 커스텀 훅으로 분리

- useUserName name을 리턴하도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: api 명세 변경에 따른 MSW 이메일 validate 로직 변경

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 모든 api call 커스텀 훅으로 분리

- accessToken 필요한 useFetch useAuthFetch로 분리
- hooks 폴더 구조화

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: 로그인 유저 , 비로그인 유저만 접근할 수 있는 path 설정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: api skip 여부도 커스텀 훅에서 정의하도록 수정

-styled-components 위치 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
feat: snackbar 적용

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
style: 장바구니 페이지 반응형 어느정도 적용

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 회원가입 enter submit 구현

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: snackbar 삭제될 때 padding도 같이 삭제해서 한번에 없어지도록 수정

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
fix: useDeleteUser에서 method: delete 추가

Co-authored-by: ronci <ronci@users.noreply.github.com>

kwannee committed:
refactor: 올바르지 않은 주소 입력 확인하는 로직 명확하도록 수정

kwannee committed:
feat: 에러 페이지 홈으로 이동할 수 있는 링크 추가

kwannee committed:
refactor: 로그인 여부에 따른 리다이렉트 해주는 컴포넌트 중복 제거

kwannee committed:
chore: 서버 api 추가

kwannee committed:
chore: 서버 url path 작동하는걸로 변경

kwannee committed:
refactor: 사용하지 않는 console.log 및 svg 삭제

EungyuCho commented on react-shopping-cart/src/components/ShoppingCartListItem/ShoppingCartListItem.component.jsx:
```diff
@@ -10,40 +10,48 @@ import TextBox from 'components/@shared/TextBox/TextBox.component';
+  const onChangeQuantity = async newQuantity => {
+    debounce(async () => {
+      await modifyCartQuantity({ productId: id, quantity: newQuantity });
+      await loadCarts();
```
음.. 수량을 올릴때 바로 결제 예상금액이 바뀌질 않으니 어색하네요 🤔 
어떤식으로 해결해볼 수 있을까요? 

저는 수량이 바뀌었다고 생각하고 가격만 업데이트해도 괜찮을 것 같다고 생각이 들어요. (주문하기라는 2차 방어선이 있어서)

etc) 만약 debounce로 서버의 부담을 덜어주려는 목적이라면 이벤트 트리거 시점에 이전 quantity에서 변화가 없다면 요청을 하지 않고 early return을 하는게 맞지 않았을까요?

kwannee replied:
일단 말씀해주신대로 수정할 방법을 생각해 봤습니다! 

현재 장바구니의 total price는 서버에서 바로 가져온 carts 중에서 현재 체크된 상품들의 price*quantity의 총합으로 계산하고 있습니다.
장바구니 수량을 수정하면 변경된 수량이 서버에 patch가 되고, 다시 carts를 get해서 변경된 carts로 새롭게 total price를 계산합니다.
그러나 현재 장바구니 수량을 수정하는 로직에 debounce를 적용해 놔서 마지막 수량 수정 이후 1000ms 이후에야 서버에 patch하고 다시 가져오는 작업을 수행합니다. 그래서 유저 입장에서는 장바구니 수량을 수정해도 total price가 즉각적으로 변경되지 않는 것 처럼 느껴집니다.
해당 사항을 유저 친화적으로 즉각적이게 변동 되도록 수정하려면 carts를 담당하는 리듀서를 새로 만들어야 할 것 같습니다. 그리고 thunk를 설치하여 비동기로 가져온 데이터도 처리할 수 있게 해야 합니다.

주절주절 말이 길었습니다. 결론적으로, thunk를 설치해서 적용해본 결과 되긴 하는데 고칠 것이 정말 많아집니다... carts와 관련된 모든 서버 api call을 하고 난 후, 변경된 데이터를 로컬의 리덕스와도 일치시키는 과정을 추가해야 했습니다. 하다보니 고쳐야 할 게 꽤나 많았고 자잘한 side effects들도 발생해서 제출 시간 내 다 수정할 수 없을 것 같다고 판단했습니다.

차선책으로 장바구니 페이지에서 상품 수량 수정 시의 디바운싱을 포기했습니다. 현재 서버에 부하가 가지 않기 때문에 굳이 디바운싱을 할 필요가 없을 것 같습니다. 
추가적으로 기존의 디바운스를 사용하는 곳에서 만약 기존과 수량이 같다면 api call은 하지 않도록 수정했습니다!

[a6848c5](https://github.com/kwannee/react-shopping-cart-prod/commit/a6848c58e0327e5d340418a4b24a0818d09b7f91)

EungyuCho replied:
좋습니다 ㅎㅎ

디바운스를 굳이 가져갈필요 없다고 생각해요. 대부분의 카트 페이지에서도 클릭시마다 업데이트 api를 호출하는 형태라 괜찮다고 생각이 드네요!

이러한 경우에 수량을 한번에 변경하고 싶은 니즈도 있을 수 있다고 생각하는데 select나 input을 함께 사용할 수 있는 구조도 고민해보면 좋을거같아요. 🙂

EungyuCho commented on react-shopping-cart/src/App.jsx:
```diff
@@ -0,0 +1,55 @@
+import GlobalStyle from 'styles/GlobalStyle';
+import theme from 'styles/theme';
+
+const AuthOnly = () => {
+  const { accessToken } = useSelector(state => state.auth);
+  return accessToken ? <Outlet /> : <Navigate to="/login" />;
+};
+const UnAuthOnly = () => {
+  const { accessToken } = useSelector(state => state.auth);
+  return !accessToken ? <Outlet /> : <Navigate to="/" />;
+};
```
👍 👍

kwannee replied:
step1에서 알려주신 덕분에 이렇게 멋진 방법을 알게 됐네요! 감사합니다!!!

EungyuCho commented:
리뷰가 조금 늦어질 것 같아서 미리 본문 코멘트부터 답변을 드릴게요!


> palette 사용법.

만약 현업에서 palette를 사용하시게 된다면 디자이너분들이 색상을 다 정의해둔 상태일꺼고 관련해서 color를 지정하는 별도 라이브러리가 있을 확률이 커요 🙂 

요구사항의 끝이 명확히 정의되어있고 구현까지 해당 색상 하나만 필요하다면 이외 색상을 지정할 필요가 없다고 생각하지만 그런 프로젝트는 거의 없으니 모든 색상을 다 정의해둘 것 같아요 ㅎㅎ

> 중첩 라우터 사용

저도 후이님의 방식이 맞다고 생각해요. 말씀해주신대로 더러워보일수는 있다고 생각하지만 확장성을 생각해봤을 때 가장 좋은 선택이 아닐까... 싶습니다 😇 

> 서버에서 오는 데이터를 사용할 수 있는 형태로 가공하는 과정이 원래 일반적인건가요?

이 부분은 정말 회바회라고 생각이 드는데요..!
서버 개발자분에 따라서 저렇게 내려주시는 분들도 계시고 반면에 프론트엔드에서 핸들링하기 편한쪽으로 작성해달라는 분들도 계시고.. 케바케입니다 💦 

백엔드에서 Product랑 수량을 관리하는 테이블이 별도로 있다보니 인터페이스 구조를 지켜주시느라 구조가 나온 것 같은데..? 저희가 잘 모르는 백엔드 크루님들의 기술적인 고민이 있었겠죠..? 🥲  

`일반적인가?` 의 관점에서는 저도 아직 주니어라 그런지 명확하게 답을 내려드리기 어렵네요 ㅎㅎ; 아직까진 반반이에요~!

요즘은... 원하시는대로 짜라고 백엔드분이 말씀해주시고 계시긴해요 😅 (행복)
만약에 이러한 부분이 해결되기 어렵다! 하면 BFF(Backend for frontend)를 도입하시는 것도 방법일 것 같아요.

몇몇 회사들은 사용하는 것 같더라구요~~ [참고링크](https://fe-developers.kakaoent.com/2022/220310-kakaopage-bff/)

EungyuCho commented on react-shopping-cart/src/components/Header/Header.component.jsx:
```diff
@@ -1,17 +1,99 @@
+  };
+
+  const handleDeleteAccount = async () => {
+    if (window.confirm(`회원 탈퇴하시겠습니까?`) && window.confirm('정말?')) {
```
2번.. 물어보네요
구현해주신 디테일이 현실적이네요 😄

kwannee replied:
하하 좀 쿨하면서도 약간은 구질구질한 느낌을 살려봤습니다.

EungyuCho commented on react-shopping-cart/src/pages/ErrorPage/ErrorPage.page.jsx:
```diff
@@ -1,12 +1,26 @@
-      <Error>{children}</Error>
+      <FlexBox direction="column" gap="50px" justifyContent="center">
+        <Error>{children}</Error>
+        <HomeLink to="/">홈으로</HomeLink>
```
컴포넌트에서 에러가 발생했을 때 홈으로 버튼을 눌러도 동작하지 않는 경우가 있었는데요. (서버에서 에러를 임의로 발생시켰을 때)
reloadDocument 속성을 켜주면 메인 페이지로 이동할 때 다시 로딩해줘서 동작을 하더라구요 🤔

kwannee replied:
오 이런게 있었군요 감사합니다! 클라이언트 측에서 라우팅 하는 것을 스킵하고 브라우저가 transition을 처리하도록 하는 속성이네요. 덕분에 하나 더 배웠습니다! 수정 완료했습니다!

[1629aa5](https://github.com/kwannee/react-shopping-cart-prod/commit/1629aa5dc2e26dff63e66e9681863f746f91917f)

EungyuCho commented on react-shopping-cart/src/pages/ErrorPage/ErrorPage.page.jsx:
```diff
@@ -1,12 +1,26 @@
+import FlexBox from 'components/@shared/FlexBox/FlexBox.component';
 import PageContainer from 'components/@shared/PageContainer/PageContainer.component';
 
 function ErrorPage({ children }) {
```
에러페이지가 범용적으로 사용되고 있는데 http status에 따라서 처리해보면 어떨까요?
가령 access token expire 되었을 때 다시 로그인을 유도할 수 있게 로그인 페이지로 리다이렉트 한다던지?..
api에서 타임아웃이 되었을 때 에러 페이지가 아니라 해당 페이지에서 버튼 하나로 api를 다시 요청할 수 있게 고쳐본다던지?
각 에러 상황마다 어떻게 유저에게 가이드라인을 주면 편할지 생각해보면 좋을거같아요~

kwannee replied:
와우 정말 중요한 관점이네요 감사합니다! 생각해보면 당연한 건데 그냥 에러 페이지 하나 만들었다는 사실에 혼자 만족 해버리고 거기서 멈춘 것 같습니다.
말씀해주신 방향으로 수정해 보려고 했으나, 조금 일이 커질 것 같다는 생각이 들었고, 현재 방학을 너무 즐겨버린 탓에 시간이 부족하여 완전히 동일한 방식으로는 수정하기는 조금 힘들 것 같습니다.ㅠㅠ
차선책으로 ErrorPage에 가는 것은 동일 하되, status에 따라 다른 처리를 하도록 수정했습니다.
error를 담당하는 리듀서를 새로 하나 만들고, 비동기 에러가 발생했을 시 status를 set 하도록 했습니다.
ErrorPage 마운트 시에는 해당 리듀서의 status를 가져와서 해당 status를 바탕으로 화면에 보일 message와 redirect 그리고 redirectMessage를 다르게 설정할 수 있도록 했습니다!

[0e38b28](https://github.com/kwannee/react-shopping-cart-prod/commit/0e38b2824dcde5dff85b35438644f88a29140695)

EungyuCho replied:
좋아요 깔끔하네요!
조금 우려되는 부분은 토큰이 만료되었을 때에도 의도하신대로 잘 working 할지 확신이 안드는데요 👀 
토큰이 잘못되었거나 만료되었을 때 어떤식으로 동작하는지 한번 확인해보시면 좋을 것 같아요.

로그인페이지로 로그인페이지로 리다이렉트를 할때에는 다시 로그인을 해야하니 관련된 키를 없애주거나 로컬스토리지를 비워보면 어떨지 생각이 들었습니당

EungyuCho reviewed (changes requested):
안녕하세요 후이님! 너무 반가워요~

배려해주신 덕분에 .env파일을 만들지 않고 간단하게 실행시킬 수 있었어요! 🙂 

이전 피드백들도 잘 반영해주신데다가 msw도 api에 맞춰서 업데이트, 비동기 모듈 분리, 코드 통일성등 칭찬드리고 싶은 부분이 많았어요 👏 

그래서 UX적인 부분이나 서버 에러 상황에 따라 예외처리하는 부분에 대해 몇가지 코멘트를 남겨드렸는데요.
확인해보시고 후이님의 생각도 들려주세요 ㅎㅎ

본문의 답변은 별도의 코멘트로 달아놨으니 참고해주시면 감사하겠습니다
방학 잘 보내세요 후이님! 🌴

kwannee committed:
fix: ErrorPage에서 홈으로 버튼 눌러도 홈으로 가지 않던 오류 reloadDocumnet 추가해서 수정

kwannee committed:
feat: 에러 종류에 따라 에러 페이지 다르게 표시

kwannee committed:
fix: debounce 하지 않고 바로 수량 업데이트 하도록 수정

-상품 목록에서 장바구니 수량 수정할 때 처음 수량과 동일하다면 api call 하지 않도록 수정
-debounce delay 너무 길게 설정 돼 있던 것 수정
-장바구니 페이지 담긴 수량 vertical center가 되도록 스타일링

kwannee commented:
안녕하세요 아사님! 후이입니다!
피드백 수정 pr이 늦어져서 죄송합니다 ㅜㅜ 방학 시작하자 마자 놀러갈 계획을 짜버려서 펑펑 놀아버렸습니다...
늦게서야 수정을 해봤는데 시간이 촉박해서 완전히 다 수정 하지는 못했습니다. 대신 비슷하게라도 수정하려고 노력해봤습니다.
피드백 주신 UX적 관점은 정말정말 중요한 것 같습니다. 꼭 유념하도록 하겠습니다!

EungyuCho commented on react-shopping-cart/src/components/ProductListItem/ProductListItem.component.jsx:
```diff
@@ -1,15 +1,118 @@
+  };
+  const handleChangeQuantity = async newQuantity => {
+    isModified.current = true;
+
+    debounce(async () => {
+      setModifyQuantityShow(false);
+
+      isModified.current = false;
+
+      if (newQuantity === quantity) {
+        return;
+      }
+
+      if (newQuantity !== 1) {
+        dispatch(setSnackBarMessage(`🛒 ${name} ${newQuantity}개가 장바구니에 담겼습니다!`));
+      }
+      await modifyCartQuantity({ productId: id, quantity: newQuantity });
+      await loadProducts();
+    }, 1500);
+  };
+
+  const handleStoreProductToCart = async () => {
+    if (!accessToken) {
+      dispatch(setSnackBarMessage('로그인 해주세요!'));
+      return;
+    }
+    handleShowModifyQuantityBox();
+
+    await storeCart({ productId: id, quantity: 1 });
+    await loadProducts();
+  };
+
+  const handleDeleteProduct = async () => {
+    dispatch(setSnackBarMessage(`🛒 ${name}가 장바구니에서 삭제됐습니다!`));
+
+    await deleteCarts({ productIds: [id] });
+    await loadProducts();
+
+    setModifyQuantityShow(false);
+  };
```
hook으로 추상화 하는편이 낫지 않았을까요? 🤔 
마크업과 세부 동작에 대한 로직이 섞여있으면 코드가 복잡하다고 느껴져서 특정 기능에 대한 추상화가 가능하다고 생각되시는 부분이 있다면 시도해보시는걸 추천드려요

EungyuCho commented on react-shopping-cart/src/pages/ShoppingCartList/ShoppingCartList.page.jsx:
```diff
@@ -13,78 +14,79 @@ import PaymentAmountContainer from 'components/PaymentAmountContainer/PaymentAmo
-  const checked = shoppingCart.length !== 0 && orderList.length === shoppingCart.length;
+  const { carts, isLoading, loadCarts } = useLoadCarts();
+  const { items: orderList } = useSelector(state => state.orderList);
+  const { deleteCarts } = useDeleteCart(true);
```
여기  인자로 넣어주신 true는 사용되는걸까요?

kwannee replied:
하이고... 처음에 skip 여부를 인자로 받았었는데 명시적이지 못하기도 하고, skip 여부가 달라질 것 같지 않아서 모두 제거해 줬는데 이녀석을 빼먹었네요... 확인 한다고 확인 했는데 이런 실수를...

EungyuCho reviewed (approved):
고생하셨습니다 후이님 :)

마지막 미션 고생 많으셨어요! 놀러 다녀오셨군요 재밌으셨나요?? 😄 (항상 뒤늦게 후회하는 부분인데 쉴 수 있을때 쉬어두는게 좋더라구요 🍾 )

최근에 있어 공유를 하고싶어 주저리주저리 적어보려고 해요.

점점 협업에 대해 집중해야하는 과제들을 만날거라고 생각이 드는데요.
단순히 단발성 프로젝트 뿐만 아니라 장기적인 프로젝트를 가져갈 때 어떻게하면 다른 팀원이 와도 유지보수하기 좋을 지 항상 고민해보시는걸 권장드려요.

예를들어 유효성 검사에 필요한 조건이 있었을때 정규식 변수를 사용한다고 치면 정규식 변수 하나만 사용하고도 간단하게 검증이 가능한데요, 요구사항이 조금 바뀐다고 쳤을 때 다른 팀원이 유지보수 하기 힘들수도 있다고 생각해요. 🤔 

그래서 선언적으로 함수를 만들어서 해결할 수 없는지 고민해보고 가능하다면 적용하는게 맞는 것 같아요. 성능에 대한 부분이 떨어지더라도 직관적인 함수를 통해 해결하는게 유지보수에 대한 리소스를 더 줄일 수 있기 때문에..! (하지만 이 부분도 프로덕트에 따라 달라질 수 있다고 생각해요 얼마나 빠르게 서빙되어야 하는지?, 자주 바뀔 여지가 없는지? 등을 고려..)

이번 미션에서 짚어드린 사용자 관점에서 어떻게 하면 유저가 더 익숙한 방식으로 풀어낼지(annoying한 동작을 유발하고 있지 않는지) 그리고 만들고 있는 제품이 처음에 설정한 원칙에 벗어나진 않았는지? 이외에도 실제로 프로덕트를 만들다 보면 개발을 하면서도 다양한 부분을 팀원들과 이러한 부분들을 함께 생각해봐야하는데요. 이런 부분도 우테코의 협업과정에서 충분히 학습하실 수 있을거라고 생각해요 🙂 (함께 자라기의 양치질 사례처럼 의도적 학습이 중요하다고 생각해요.)

아무래도 마지막 피드백이라고 생각하니 이런 저런 생각이 많이 드네요 😢 
막상 마지막이 되니 크루들에게 조금 더 용기내서 다가가볼껄 하는 후회가 많이 밀려오네요 ㅎㅎ;;

여튼 이번 미션 깔끔하게 구현해주셔서 좋았어요!
다음번에 만나게되면 이야기 많이 나눠보면 좋겠어요 ㅎㅎ

후이님의 앞으로의 성장도 응원할게요 👍 👍 

P.S. 크루님들 DM 언제줘요..?
